tools/update_es_based_on_url_keyword.py

```diff
@@ -0,0 +1,172 @@
+    ArticleUrlKeywordMap,
+    UrlKeyword,
+    NewArticleTopicMap,
+    NewTopic,
+)
+from zeeguu.core.model.article import article_topic_map
+from zeeguu.core.elastic.settings import ES_ZINDEX, ES_CONN_STRING
+import numpy as np
+from tqdm import tqdm
+
+app = create_app()
+app.app_context().push()
+
+
+es = Elasticsearch(ES_CONN_STRING)
+db_session = zeeguu.core.model.db.session
+print(es.info())
+
+
+def main():
+    def fetch_articles_by_id(id_list, recalculate_topic=False):
+        for i in id_list:
+            try:
+                article = Article.find_by_id(i)
+                if not article:
+                    print(f"Skipped for: '{i}', article not in DB.")
+                    continue
+                if recalculate_topic:
+                    article.recalculate_topics_from_url_keywords(db_session)
+                    db_session.commit()
+                yield (article)
+            except NoResultFound:
+                print(f"fail for: '{i}'")
+            except Exception as e:
+                print(f"fail for: '{i}', {e}")
+
+    def gen_docs(articles_w_topics):
+        for article in articles_w_topics:
+            try:
+                yield create_or_update_bulk_docs(article, db_session)
+            except Exception as e:
+                print(f"fail for: '{article.id}', {e}")
+
+    # Get the articles for the url_keyword
+    target_ids = np.array(
+        [
+            a_id[0]
+            for a_id in db_session.query(Article.id)
+            .join(ArticleUrlKeywordMap)
+            .join(UrlKeyword)
+            .filter(UrlKeyword.keyword == URL_KEYWORD_TO_UPDATE)
+            .distinct()
+        ]
+    )
+
+    print(
+        f"Got articles with url_keyword '{URL_KEYWORD_TO_UPDATE}', total: {len(target_ids)}",
+    )
+
+    # Updating url_keyword new_topic mapping
+    # And the topics that were added based on that keyword.
+    if DELETE_ARTICLE_NEW_TOPICS:
+        new_topics_ids_to_delete = []
+        new_topics = []
+        url_keywords = UrlKeyword.find_all_by_keyword(URL_KEYWORD_TO_UPDATE)
+        for u_key in url_keywords:
+            if u_key.new_topic_id:
+                new_topics.append(NewTopic.find_by_id(u_key.new_topic_id))
+                new_topics_ids_to_delete.append(u_key.new_topic_id)
+                u_key.new_topic_id = None
+
+        print(
+            f"Deleting new_topics '{",".join([t.title for t in new_topics])}' for articles which have the keyword: '{URL_KEYWORD_TO_UPDATE}'"
+        )
+        topic_mappings_to_delete = NewArticleTopicMap.query.filter(
+            NewArticleTopicMap.article_id.in_(list(target_ids))
+        ).filter(NewArticleTopicMap.new_topic_id.in_(new_topics_ids_to_delete))
+        print(
+            f"Found '{len(topic_mappings_to_delete.all())}' topic mappings to delete."
+        )
+        topic_mappings_to_delete.delete()
+        db_session.commit()
+        print("MySQL deletion completed.")
+
+    if len(target_ids) == 0:
```

Comment:
should this be moved above, immediately after the getting of the `target_id`s? the `print('Got articles...` is on the other branch of this if I think.

Reply:
I don't think so. This has to do with re-indexing, not the data integrity. If I moved this above, then there would be data that wouldn't be updated in the database, which I think is more confusing. 

The way I see it, is that the DB should always be consistent regarding the mappings, but the ES can be incomplete, e.g. some articles might not be in ES and we do not necessarily need to index everything in ES.

Reply:
With some testing, I realized that originally, I was only updating the articles in the DB that were in ES. I still think we should keep the DB data correct with the updates, even if it takes somewhat longer - but ES can be a bit more flexible. This was related with the 'fetch_articles_by_id' not being clean, and re-calcualting the topics when fetching. 

I think the flow now makes more sense: Delete Mappings and the Associating of the Keyword -> Recalculate the Topics for Articles -> Index them in ES.